samza-core/src/main/java/org/apache/samza/zk/ZkJobCoordinator.java

```diff
@@ -0,0 +1,221 @@
+import org.apache.samza.system.SystemAdmin;
+import org.apache.samza.system.SystemFactory;
+import org.apache.samza.util.SystemClock;
+import org.apache.samza.util.Util;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * JobCoordinator for stand alone processor managed via Zookeeper.
+ */
+public class ZkJobCoordinator implements JobCoordinator, ZkControllerListener {
+  private static final Logger log = LoggerFactory.getLogger(ZkJobCoordinator.class);
+
+  private final ZkUtils zkUtils;
+  private final int processorId;
+  private final ZkController zkController;
+  private final SamzaContainerController containerController;
+  private final BarrierForVersionUpgrade barrier;
+  private final ScheduleAfterDebounceTime debounceTimer;
+  private final StreamMetadataCache  streamMetadataCache;
+  private final ZkKeyBuilder keyBuilder;
+  private final Config config;
+
+  private JobModel newJobModel;
+  private String newJobModelVersion;  // version published in ZK (by the leader)
+  private JobModelManager jobModelManager;
+
+  public ZkJobCoordinator(int processorId, Config config, ScheduleAfterDebounceTime debounceTimer, ZkUtils zkUtils, SamzaContainerController containerController) {
+    this.zkUtils = zkUtils;
+    this.keyBuilder = zkUtils.getKeyBuilder();
+    this.debounceTimer = debounceTimer;
+    this.processorId = processorId;
+    this.containerController = containerController;
+    this.zkController = new ZkControllerImpl(String.valueOf(processorId), zkUtils, debounceTimer, this);
+    this.config = config;
+
+
+    barrier = new ZkBarrierForVersionUpgrade(zkUtils, debounceTimer);
+    streamMetadataCache = initJobModelGeneration();
+  }
+
+  private StreamMetadataCache initJobModelGeneration() {
+    // model generation - NEEDS TO BE REVIEWED
+    JavaSystemConfig systemConfig = new JavaSystemConfig(this.config);
+    Map<String, SystemAdmin> systemAdmins = new HashMap<>();
+    for (String systemName: systemConfig.getSystemNames()) {
+      String systemFactoryClassName = systemConfig.getSystemFactory(systemName);
+      if (systemFactoryClassName == null) {
+        String msg = String.format("A stream uses system %s, which is missing from the configuration.", systemName);
+        log.error(String.format(msg));
+        throw new SamzaException(msg);
+      }
+      SystemFactory systemFactory = Util.getObj(systemFactoryClassName);
+      systemAdmins.put(systemName, systemFactory.getAdmin(systemName, this.config));
+    }
+
+    return new StreamMetadataCache(Util.<String, SystemAdmin>javaMapAsScalaMap(systemAdmins), 5000, SystemClock.instance());
+  }
+
+  @Override
+  public void start() {
+    zkController.register();
+  }
+
+  public void cleanupZk() {
+    zkUtils.deleteRoot();
+  }
+
+  @Override
+  public void stop() {
+    zkController.stop();
+  }
+
+  @Override
+  public boolean awaitStart(long timeoutMs)
+      throws InterruptedException {
+    return containerController.awaitStart(timeoutMs);
```

Comment:
The awaitStart behavior is also confusing. At this point, if containerController.startContainer has not been called, then awaitStart will fail since the container has not yet been initialized. We need to fix this by de-coupling the Streamprocessor API from the JobCoordinator API. I created [SAMZA-1147](https://issues.apache.org/jira/browse/SAMZA-1147) to fix this. But even without it, this statement will fail if container has not yet been initialized.